aip/client-libraries/4210.md

```diff
@@ -31,25 +35,95 @@ expectation.
+    rank = same;
+    style = filled;
+    fillcolor = lightgrey;
+    node [ shape=rectangle ];
+
+
+    protobuf_output [ label="Message & Enum\nCode" fillcolor=deepskyblue3 ];
+    grpc_output [ label="Server & Client\nStubs" fillcolor=gold3 ];
+    gapic_output [ label="Google API\nClient" fillcolor=darkseagreen ];
+
+  }
+  
+  protobuf -> protobuf_output;
+  grpc -> grpc_output;
+  gapic -> gapic_output;
+
+  assembly [ label="Package\nassembly" shape=oval fillcolor=aliceblue ];
+  
+  protobuf_output -> assembly
+  grpc_output -> assembly
+  gapic_output -> assembly
+
+  assembled_package [ label="Package of\ngenerated code" fillcolor=aliceblue ];
+
+  assembly -> assembled_package
+}
+```
+
+The following sections focus on the "GAPIC generator" in the above diagram.
+
 ### Protobuf plugins
 
-Code generators **must** be implemented as plugins to `protoc`, the protocol
-buffer compiler. The [plugin system][0] allows plugins to be written in any
-language, and plugins **should** ordinarily be written in the language being
-targeted, in order to take advantage of in-language tooling, and to ensure that
-experts in the target environment are able to meaningfully contribute.
+The protobuf compiler, `protoc`, supports a [plugin system][0] for code
+generation. The plugin system allows plugins to be written _in_ and _for_ any
+language.
+
+Code generators **must** be implemented as a `protoc` plugin. The following
+rules apply to the implementation of a client library generator as a `protoc`
+plugin:
 
+- The plugin  **should** be written in the language being targeted for
+  generation.
 - `protoc` expects plugins to be an executable in `$PATH`, and named
   `protoc-gen-{plugin_name}`, corresponding to the `--{plugin_name}_out` option
-  sent to the `protoc` executable.
-  - For a plugin creating client libraries for a specific language, the option
-    name **should** follow the convention `--{lang}_gapic_out` (meaning the
-    corresponding plugin executable is named `protoc-gen-{lang}_gapic`).
-- Plugins **must** accept a serialized `CodeGeneratorRequest` object (defined
+  sent to the `protoc` executable. As such:
+  - the plugin executable **should** be named `protoc-gen-{lang}_gapic`
+  - the plugin option  **should** follow the convention `--{lang}_gapic_out`
+- The plugin **must** accept a serialized `CodeGeneratorRequest` object (defined
   in [`plugin.proto`][1]) on `stdin`; the bulk of this is a series of
   `FileDescriptorProto` messages (defined in [`descriptor.proto`][2]).
-- Plugins **must** emit a serialized `CodeGeneratorResponse` object (defined in
-  [`plugin.proto`][1]) on `stdout`.
+- The plugin **must** emit a serialized `CodeGeneratorResponse` object (defined
+  in [`plugin.proto`][1]) on `stdout`.
+- The plugin **must not** leverage `protoc` "insertion points". Despite the
+  `protoc` plugin documentation indicating the existence of insertion points,
+  its use is unsupported and discouraged by the Protobuf team.
```

Comment:
```suggestion
  their use is unsupported and discouraged by the Protobuf team.
```

Reply:
Took your suggestion in my own commit. Done.